benchexec/test_runexecutor.py

```diff
@@ -1199,6 +1199,109 @@ def test_uptime_without_lxcfs(self):
             uptime, 10, f"Uptime {uptime}s unexpectedly low in container"
         )
 
+    def test_fuse_overlay(self):
```

Comment:
Looks like benchexec would hang when overlaying `/tmp`:

```
benchexec-venvhaoranyang@haoranyang-fudan:~/benchexec$ containerexec --overlay-dir /tmp/ --debug -- bash
2024-08-28 10:13:26 - DEBUG - This is containerexec 3.25-dev.
2024-08-28 10:13:26 - INFO - Starting command bash
2024-08-28 10:13:26 - DEBUG - Available Cgroups: {}
2024-08-28 10:13:26 - DEBUG - Starting process.
2024-08-28 10:13:26 - DEBUG - Parent: child process of RunExecutor with PID 33063 started.
2024-08-28 10:13:26 - DEBUG - Child: child process of RunExecutor with PID 33063 started
2024-08-28 10:13:26 - DEBUG - Skipping directory mount for b'/tmp' because parent already has same mode.
2024-08-28 10:13:26 - DEBUG - fuse-overlayfs version: 1.13
2024-08-28 10:13:26 - DEBUG - Creating overlay mount with /usr/bin/fuse-overlayfs: target=b'/tmp/BenchExec_run_hhj0ihsv/temp/fuse', lower=b'/', upper=b'/tmp/BenchExec_run_hhj0ihsv/temp', work=b'/tmp/BenchExec_run_hhj0ihsv/overlayfs/fuse'
2024-08-28 10:13:26 - DEBUG - Mounting '/' as overlay
2024-08-28 10:13:26 - DEBUG - [Errno 16] umount(b'/tmp/BenchExec_run_hhj0ihsv/mount/') failed: Device or resource busy
2024-08-28 10:13:26 - DEBUG - Mounting '/dev' as read-only
2024-08-28 10:13:26 - DEBUG - Mounting '/dev/pts' as read-only
2024-08-28 10:13:26 - DEBUG - Mounting '/dev/shm' as read-only
2024-08-28 10:13:26 - DEBUG - Mounting '/dev/mqueue' as read-only
2024-08-28 10:13:26 - DEBUG - Mounting '/dev/hugepages' as read-only
2024-08-28 10:13:26 - DEBUG - Mounting '/run' as hidden
2024-08-28 10:13:26 - DEBUG - Mounting '/sys' as read-only
2024-08-28 10:13:26 - DEBUG - Mounting '/sys/kernel/security' as read-only
2024-08-28 10:13:26 - DEBUG - Mounting '/sys/fs/cgroup' as read-only
2024-08-28 10:13:26 - DEBUG - Mounting '/sys/fs/pstore' as read-only
2024-08-28 10:13:26 - DEBUG - Mounting '/sys/fs/bpf' as read-only
2024-08-28 10:13:26 - DEBUG - Mounting '/sys/kernel/tracing' as read-only
2024-08-28 10:13:26 - DEBUG - Mounting '/sys/kernel/debug' as read-only
2024-08-28 10:13:26 - DEBUG - Mounting '/sys/fs/fuse/connections' as read-only
2024-08-28 10:13:26 - DEBUG - Mounting '/sys/kernel/config' as read-only
2024-08-28 10:13:26 - DEBUG - Mounting '/proc' as read-only
2024-08-28 10:13:26 - DEBUG - Mounting '/boot' as overlay
2024-08-28 10:13:26 - DEBUG - Mounting '/var' as overlay
2024-08-28 10:13:26 - DEBUG - Mounting '/var/lib/lxcfs' as read-only
2024-08-28 10:13:26 - DEBUG - Mounting '/snap/snapd/21759' as overlay
2024-08-28 10:13:26 - DEBUG - Mounting '/snap/snapd/21465' as overlay
2024-08-28 10:13:26 - DEBUG - Mounting '/snap/core22/1439' as overlay
2024-08-28 10:13:26 - DEBUG - Mounting '/snap/lxd/29943' as overlay
2024-08-28 10:13:26 - DEBUG - Mounting '/snap/lxd/29727' as overlay
2024-08-28 10:13:26 - DEBUG - Mounting '/snap/core22/1564' as overlay
2024-08-28 10:13:26 - DEBUG - Mounting '/tmp/BenchExec_run_hhj0ihsv' as overlay
```

Reply:
Interesting. This doesn't occur with the current main branch. For the test it would probably be enough to use a subdirectory of `/tmp` instead of `/tmp` directly. But it might be good to investigate what is happening here such that BenchExec does not break if a user passes `--overlay-dir /tmp`.

Reply:
Ok, a test on the current state of the branch shows that the following hangs:
```
containerexec --overlay-dir /tmp --debug bash
```
Looks similar to the bad case with old fuse-overlayfs but also happens with fuse-overlayfs 1.13. The last line of the log is
```
2024-09-13 14:05:04 - DEBUG - Mounting '/tmp/BenchExec_run_734cy4ft' as overlay
```

I think we should simply always enforce that the temp directory of BenchExec is mounted in hidden mode, that should solve this problem as well.

The following command has the same problem:
```
containerexec --keep-tmp --debug bash
```
Here, `/tmp/BenchExec_run_...` is attempted as a full-access mount. But I guess it will also just work after the same fix, just mentioning this command line for testing.

Reply:
Some more discussion is in the comments starting with this one: https://github.com/sosy-lab/benchexec/pull/1062#issuecomment-2354889831

We did not find a solution and decided to simply declare this combination as unsupported. A detection of the case and an error message that prevents the deadlock were added in 33249f172beb6a7f7b73b16b894d14d9dfa5ad2a.